File: apps/csharp/tests/Ch02/PasswordCrackerTests.cs
using GrokkingConcurrency.CSharp.Ch02;

namespace GrokkingConcurrency.CSharp.Tests.Ch02;

/// <summary>
/// パスワードクラッカーのテスト（逐次処理版）
/// </summary>
public class PasswordCrackerTests
{
    [Fact]
    public void ShouldGenerateFourDigitCombinations()
    {
        var combinations = PasswordCracker.GetCombinations(4, 0, 9);

        Assert.Equal(10, combinations.Count);
        Assert.Equal(
            new[] { "0000", "0001", "0002", "0003", "0004", "0005", "0006", "0007", "0008", "0009" },
            combinations
        );
    }

    [Fact]
    public void ShouldGenerateCombinationsInRange()
    {
        var combinations = PasswordCracker.GetCombinations(4, 98, 102);

        Assert.Equal(5, combinations.Count);
        Assert.Equal(new[] { "0098", "0099", "0100", "0101", "0102" }, combinations);
    }

    [Fact]
    public void ShouldCalculateSha256Hash()
    {
        var hash = PasswordCracker.GetCryptoHash("12345678");

        Assert.Equal("ef797c8118f02dfb649607dd5d3f8c7623048c9c063d532cc95c5ed7a898a64f", hash);
    }

    [Fact]
    public void ShouldCheckPasswordMatch()
    {
        var expectedHash = "ef797c8118f02dfb649607dd5d3f8c7623048c9c063d532cc95c5ed7a898a64f";

        Assert.True(PasswordCracker.CheckPassword(expectedHash, "12345678"));
        Assert.False(PasswordCracker.CheckPassword(expectedHash, "87654321"));
    }

    [Fact]
    public void ShouldCrackPassword()
    {
        // "1234" の SHA-256 ハッシュ
        var cryptoHash = PasswordCracker.GetCryptoHash("1234");

        var result = PasswordCracker.CrackPassword(cryptoHash, 4);

        Assert.Equal("1234", result);
    }

    [Fact]
    public void ShouldReturnNullWhenNotFound()
    {
        // 存在しないハッシュ
        var cryptoHash = "invalid_hash";

        var result = PasswordCracker.CrackPassword(cryptoHash, 2);

        Assert.Null(result);
    }
}

File: apps/fsharp/tests/PasswordCrackerTests.fs
module PasswordCrackerTests

open Xunit
open GrokkingConcurrency.Ch02.PasswordCracker

[<Fact>]
let ``getCombinations generates all combinations of given length`` () =
    let combinations = getCombinations 2
    Assert.Contains("aa", combinations)
    Assert.Contains("az", combinations)
    Assert.Contains("za", combinations)
    Assert.Contains("zz", combinations)
    Assert.Equal(676, List.length combinations) // 26 * 26

[<Fact>]
let ``getCryptoHash returns SHA-256 hash`` () =
    let hash = getCryptoHash "ab"
    Assert.Equal("fb8e20fc2e4c3f248c60c39bd652f3c1347298bb977b8b4d5903b85055620603", hash)

[<Fact>]
let ``checkPassword returns true for matching hash`` () =
    let hash = getCryptoHash "test"
    Assert.True(checkPassword "test" hash)

[<Fact>]
let ``checkPassword returns false for non-matching hash`` () =
    let hash = getCryptoHash "test"
    Assert.False(checkPassword "wrong" hash)

[<Fact>]
let ``crackPassword finds password for given hash`` () =
    let password = "ab"
    let hash = getCryptoHash password
    Assert.Equal(Some password, crackPassword hash 2)

[<Fact>]
let ``crackPassword returns None for non-existent password`` () =
    let hash = "0000000000000000000000000000000000000000000000000000000000000000"
    Assert.Equal(None, crackPassword hash 2)
